File: vortexdbg-android/src/main/java/com/vortexdbg/linux/android/AndroidARM64Emulator.java
package com.vortexdbg.linux.android;

import com.vortexdbg.AndroidEmulator;
import com.vortexdbg.Family;
import com.vortexdbg.arm.AbstractARM64Emulator;
import com.vortexdbg.arm.backend.BackendFactory;
import com.vortexdbg.file.FileSystem;
import com.vortexdbg.file.linux.AndroidFileIO;
import com.vortexdbg.file.linux.LinuxFileSystem;
import com.vortexdbg.linux.ARM64SyscallHandler;
import com.vortexdbg.linux.AndroidElfLoader;
import com.vortexdbg.linux.android.dvm.DalvikVM64;
import com.vortexdbg.linux.android.dvm.VM;
import com.vortexdbg.memory.Memory;
import com.vortexdbg.memory.SvcMemory;
import com.vortexdbg.spi.Dlfcn;
import com.vortexdbg.spi.LibraryFile;
import com.vortexdbg.unix.UnixSyscallHandler;
import com.vortexdbg.unwind.Unwinder;

import java.io.File;
import java.net.URL;
import java.util.Collection;

/**
 * android arm emulator
 * Created by zhkl0228 on 2017/5/2.
 */

public class AndroidARM64Emulator extends AbstractARM64Emulator<AndroidFileIO> implements AndroidEmulator {

    protected AndroidARM64Emulator(String processName, File rootDir, Collection<BackendFactory> backendFactories) {
        super(processName, rootDir, Family.Android64, backendFactories);
    }

    @Override
    protected FileSystem<AndroidFileIO> createFileSystem(File rootDir) {
        return new LinuxFileSystem(this, rootDir);
    }

    @Override
    protected Memory createMemory(UnixSyscallHandler<AndroidFileIO> syscallHandler, String[] envs) {
        return new AndroidElfLoader(this, syscallHandler);
    }

    @Override
    protected Dlfcn createDyld(SvcMemory svcMemory) {
        return new ArmLD64(backend, svcMemory);
    }

    @Override
    protected UnixSyscallHandler<AndroidFileIO> createSyscallHandler(SvcMemory svcMemory) {
        return new ARM64SyscallHandler(svcMemory);
    }

    private VM createDalvikVMInternal(File apkFile) {
        return new DalvikVM64(this, apkFile);
    }

    @Override
    public LibraryFile createURLibraryFile(URL url, String libName) {
        return new URLibraryFile(url, libName, -1, true);
    }

    @Override
    protected boolean isPaddingArgument() {
        return false;
    }

    private VM vm;

    @Override
    public VM createDalvikVM() {
        return createDalvikVM((File) null);
    }

    @Override
    public final VM createDalvikVM(File apkFile) {
        if (vm != null) {
            throw new IllegalStateException("vm is already created");
        }
        vm = createDalvikVMInternal(apkFile);
        return vm;
    }

    @Override
    public VM createDalvikVM(Class<?> callingClass) {
        return createDalvikVM(new File(callingClass.getProtectionDomain().getCodeSource().getLocation().getPath()));
    }

    @Override
    public final VM getDalvikVM() {
        return vm;
    }

    @Override
    public Unwinder getUnwinder() {
        return new AndroidARM64Unwinder(this);
    }
}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/android/AndroidARM64Emulator.kt
package com.vortexdbg.linux.android

import com.vortexdbg.AndroidEmulator
import com.vortexdbg.Family
import com.vortexdbg.arm.AbstractARM64Emulator
import com.vortexdbg.arm.backend.BackendFactory
import com.vortexdbg.file.FileSystem
import com.vortexdbg.file.linux.AndroidFileIO
import com.vortexdbg.file.linux.LinuxFileSystem
import com.vortexdbg.linux.ARM64SyscallHandler
import com.vortexdbg.linux.AndroidElfLoader
import com.vortexdbg.linux.android.dvm.DalvikVM64
import com.vortexdbg.linux.android.dvm.VM
import com.vortexdbg.memory.Memory
import com.vortexdbg.memory.SvcMemory
import com.vortexdbg.spi.Dlfcn
import com.vortexdbg.spi.LibraryFile
import com.vortexdbg.unix.UnixSyscallHandler
import com.vortexdbg.unwind.Unwinder

import java.io.File
import java.net.URL

/** ARM64 (AArch64) Android emulator: wires up the loader, syscall handler, dyld and Dalvik VM. */
open class AndroidARM64Emulator internal constructor(
    processName: String?,
    rootDir: File?,
    backendFactories: Collection<BackendFactory>?
) : AbstractARM64Emulator<AndroidFileIO>(processName, rootDir, Family.Android64, backendFactories), AndroidEmulator {

    override fun createFileSystem(rootDir: File): FileSystem<AndroidFileIO> {
        return LinuxFileSystem(this, rootDir)
    }

    override fun createMemory(syscallHandler: UnixSyscallHandler<AndroidFileIO>, envs: Array<String>): Memory {
        return AndroidElfLoader(this, syscallHandler)
    }

    override fun createDyld(svcMemory: SvcMemory): Dlfcn {
        return ArmLD64(backend, svcMemory)
    }

    override fun createSyscallHandler(svcMemory: SvcMemory): UnixSyscallHandler<AndroidFileIO> {
        return ARM64SyscallHandler(svcMemory)
    }

    private fun createDalvikVMInternal(apkFile: File?): VM {
        return DalvikVM64(this, apkFile)
    }

    override fun createURLibraryFile(url: URL, libName: String): LibraryFile {
        return URLibraryFile(url, libName, -1, true)
    }

    override fun isPaddingArgument(): Boolean {
        return false
    }

    private var vm: VM? = null

    override fun createDalvikVM(): VM {
        return createDalvikVM(null as File?)
    }

    final override fun createDalvikVM(apkFile: File?): VM {
        if (vm != null) {
            throw IllegalStateException("vm is already created")
        }
        vm = createDalvikVMInternal(apkFile)
        return vm!!
    }

    override fun createDalvikVM(callingClass: Class<*>): VM {
        return createDalvikVM(File(callingClass.protectionDomain.codeSource.location.path))
    }

    final override fun getDalvikVM(): VM? {
        return vm
    }

    override fun getUnwinder(): Unwinder {
        return AndroidARM64Unwinder(this)
    }
}
